File: benchmarks/benchkit/memory.py
"""Memory instrumentation: what a solve actually holds, on the host and the device.

The counterpart of ``src/memory.jl``, and deliberately built the same way.

Host memory is the process resident set, sampled by a background thread. That is
the quantity ``src/memory.jl`` samples too, which is what makes a Julia figure
and a Python figure comparable at all — and it is the only choice that covers
PuMA, whose solver allocates in C where a Python-level tracer such as
``tracemalloc`` sees nothing.

Device memory does not need sampling on this side: PyTorch's caching allocator
keeps an exact high-water mark, so the peak is read rather than estimated.
``max_memory_allocated`` is what a solve holds and ``max_memory_reserved`` is
what the allocator took from the driver; the second is recorded for context and
must not be compared between configurations, because a caching allocator's
footprint is dominated by its history.

``torch`` is imported lazily on purpose. Importing it alongside ``pumapy`` in one
process aborts on a duplicate OpenMP runtime, which is one of several reasons
every tool gets its own process.
"""

import threading
import time
from dataclasses import dataclass
from typing import Any

import psutil

_PROCESS = psutil.Process()


def current_rss():
    """Resident set of this process, in bytes.

    The right quantity for "how much memory did this need": it counts what is
    actually resident, includes allocations made by C libraries, and — unlike a
    peak counter — can be sampled repeatedly.
    """
    return int(_PROCESS.memory_info().rss)


def _torch():
    import torch

    return torch


def reset_device_peaks(device):
    """Clear the allocator's high-water marks so the next reading is this solve's."""
    if str(device).startswith("cuda"):
        torch = _torch()
        torch.cuda.synchronize()
        torch.cuda.empty_cache()
        torch.cuda.reset_peak_memory_stats()


def device_peaks(device):
    """``(allocated, reserved)`` peak device bytes since the last reset."""
    if not str(device).startswith("cuda"):
        return 0, 0
    torch = _torch()
    torch.cuda.synchronize()
    return int(torch.cuda.max_memory_allocated()), int(torch.cuda.max_memory_reserved())


def release_device(device):
    """Return cached blocks to the driver between cases."""
    if str(device).startswith("cuda"):
        torch = _torch()
        torch.cuda.synchronize()
        torch.cuda.empty_cache()


@dataclass
class PeakUsage:
    """What a sampled run held, alongside whatever the measured code returned."""

    value: Any
    elapsed: float
    baseline_rss: int
    peak_rss: int
    peak_device: int
    pool_device: int
    samples: int


class _RssSampler(threading.Thread):
    """Track the largest resident set seen while the measured code runs."""

    def __init__(self, interval_s):
        super().__init__(daemon=True)
        self.interval_s = interval_s
        self.peak = current_rss()
        self.samples = 0
        # Not `_stop`: `threading.Thread` already uses that name for an internal
        # method that `join` calls, and shadowing it makes every join raise.
        self._finished = threading.Event()

    def run(self):
        while not self._finished.is_set():
            self.sample()
            self._finished.wait(self.interval_s)

    def sample(self):
        self.peak = max(self.peak, current_rss())
        self.samples += 1

    def stop(self):
        self._finished.set()
        self.join(timeout=5.0)


def with_peak_sampling(fn, *, interval_ms=10, device="cpu"):
    """Run ``fn`` while memory is monitored, and report the peaks.

    The sampler thread keeps running through the measured call because the
    solvers spend their time in C or CUDA code that releases the interpreter
    lock. A final sample is taken with everything ``fn`` allocated still
    reachable, so the end state is represented even if the thread was never
    scheduled.
    """
    import gc

    gc.collect()
    reset_device_peaks(device)

    sampler = _RssSampler(interval_ms / 1000.0)
    baseline = current_rss()
    sampler.start()
    try:
        started = time.perf_counter()
        value = fn()
        if str(device).startswith("cuda"):
            _torch().cuda.synchronize()
        elapsed = time.perf_counter() - started
        sampler.sample()
    finally:
        sampler.stop()

    peak_device, pool_device = device_peaks(device)
    return PeakUsage(value, elapsed, baseline, sampler.peak, peak_device, pool_device, sampler.samples)

File: benchmarks/src/memory.jl
# Memory instrumentation: what a solve actually holds, on the host and the device.
#
# Two decisions shape this file.
#
# First, memory is measured by its own stage and never read off a timing sweep.
# The two questions want opposite things from a run — a timing must not be
# perturbed and so cannot afford a sampler, while a peak needs one and does not
# care what it costs. Separating them is what lets each be measured properly
# instead of both being measured badly.
#
# Second, the host figure is the process resident set, sampled, in both
# languages. `benchkit/memory.py` samples the same operating-system quantity
# through `psutil`, which is what makes a Julia number and a Python number
# comparable at all. Julia's own `Base.gc_live_bytes` would count only the parts
# of a solve that Julia's collector manages, and PuMA's solver is C.

using CUDA

const PAGESIZE = Ref{Int}(0)

"""Resident set of this process, in bytes.

The resident set is the right quantity for "how much memory did this need": it
counts what is actually in physical memory, includes allocations made by C
libraries, and — unlike a peak counter — can be sampled repeatedly.

Read straight from the OS on Linux and Windows. On any other platform this falls
back to `Sys.maxrss`, which is a high-water mark that never falls, so a sampled
peak there degrades to "the largest this process ever was".
"""
function current_rss()
    @static if Sys.islinux()
        return _rss_linux()
    elseif Sys.iswindows()
        return _rss_windows()
    else
        return Int(Sys.maxrss())
    end
end

function _rss_linux()
    PAGESIZE[] == 0 && (PAGESIZE[] = Int(ccall(:getpagesize, Cint, ())))
    fields = split(read("/proc/self/statm", String))
    length(fields) >= 2 || return Int(Sys.maxrss())
    return parse(Int, fields[2]) * PAGESIZE[]
end

# PROCESS_MEMORY_COUNTERS on 64-bit Windows: two DWORDs then eight SIZE_Ts, so
# 72 bytes with WorkingSetSize — the resident set — at offset 16.
const _PMC_BYTES = 72
const _PMC_WORKINGSET_OFFSET = 16

function _rss_windows()
    buf = zeros(UInt8, _PMC_BYTES)
    handle = ccall((:GetCurrentProcess, "kernel32"), Ptr{Cvoid}, ())
    rss = GC.@preserve buf begin
        p = pointer(buf)
        unsafe_store!(Ptr{UInt32}(p), UInt32(_PMC_BYTES))
        ok = ccall((:GetProcessMemoryInfo, "psapi"), Cint,
                   (Ptr{Cvoid}, Ptr{UInt8}, UInt32), handle, p, UInt32(_PMC_BYTES))
        ok == 0 ? -1 : Int(unsafe_load(Ptr{UInt64}(p + _PMC_WORKINGSET_OFFSET)))
    end
    return rss < 0 ? Int(Sys.maxrss()) : rss
end

"""Device bytes actually allocated, excluding blocks the pool merely caches.

This is the figure that answers "how much device memory does this solve need".
Its counterpart [`device_pool_bytes`](@ref) is recorded alongside for context but
must not be compared between configurations: an idle session can show tens of
gigabytes of pool against half a megabyte live, and under pressure the pool
saturates at the card's capacity and reports the same number for every
configuration — precisely where the comparison matters most.
"""
device_live_bytes() = CUDA.functional() ? Int(CUDA.memory_stats().live) : 0

"""Device bytes the driver reports as unavailable: CUDA.jl's pool footprint."""
device_pool_bytes() = CUDA.functional() ? Int(CUDA.total_memory() - CUDA.available_memory()) : 0

"""What a sampled run held, alongside whatever the measured code returned."""
struct PeakUsage
    value::Any
    elapsed::Float64
    baseline_rss::Int
    peak_rss::Int
    peak_device::Int
    peak_pool::Int
    samples::Int
end

"""Run `f` while a background task samples memory, and report the peaks.

The sampler is placed on Julia's interactive thread pool when one exists, so it
keeps sampling while the measured code saturates the default pool. Start the
process with `-t <n>,1` to get one. With no spare thread the peaks degrade to
readings taken either side of `f`, which for a Krylov solve — whose workspace is
allocated up front and then reused — is close but not guaranteed to catch a
transient during setup.
"""
function with_peak_sampling(f; interval_ms::Real=10, gpu::Bool=false)
    GC.gc(true)
    gpu && CUDA.functional() && CUDA.reclaim()

    baseline_rss = current_rss()
    peak_rss = Threads.Atomic{Int}(baseline_rss)
    peak_device = Threads.Atomic{Int}(gpu ? device_live_bytes() : 0)
    peak_pool = Threads.Atomic{Int}(gpu ? device_pool_bytes() : 0)
    samples = Threads.Atomic{Int}(0)
    running = Threads.Atomic{Bool}(true)

    take_sample!() = begin
        Threads.atomic_max!(peak_rss, current_rss())
        if gpu
            Threads.atomic_max!(peak_device, device_live_bytes())
            Threads.atomic_max!(peak_pool, device_pool_bytes())
        end
        Threads.atomic_add!(samples, 1)
    end

    sampler = _spawn_sampler(running, take_sample!, interval_ms / 1000)
    local value, elapsed
    try
        elapsed = @elapsed (value = f())
        # A last reading with everything the solve allocated still reachable: the
        # caller drops its references only after this returns, so the end state is
        # always represented even if the sampler was never scheduled.
        take_sample!()
    finally
        running[] = false
        sampler === nothing || wait(sampler)
    end

    return PeakUsage(value, elapsed, baseline_rss, peak_rss[], peak_device[],
                     peak_pool[], samples[])
end

function _spawn_sampler(running, take_sample!, interval_s)
    loop() = begin
        while running[]
            take_sample!()
            sleep(interval_s)
        end
    end
    if Threads.nthreads(:interactive) >= 1
        return Threads.@spawn :interactive loop()
    elseif Threads.nthreads() >= 2
        return Threads.@spawn loop()
    end
    @warn "no spare thread for the memory sampler — peaks fall back to readings either side " *
          "of the solve; start Julia with `-t $(Threads.nthreads()),1`" maxlog = 1
    return nothing
end
